### PR TITLE
chore(pyroscope): Ensure presence of service_name label

### PR DESCRIPTION
**Pull request by korniltsev**
#### PR Description

The special label `service_name` is required and must always be present. If it's not specified, it is
attempted to be inferred from multiple sources: 
- `__meta_kubernetes_pod_annotation_pyroscope_io_service_name` which is a `pyroscope.io/service_name` pod annotation.
- `__meta_kubernetes_namespace` and `__meta_kubernetes_pod_container_name`
- `__meta_docker_container_name`

If `service_name` is not specified and could not be inferred it is set to `unspecified`.

#### Which issue(s) this PR fixes

https://github.com/grafana/phlare/issues/711

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: component/pyroscope/scrape/target.go
Comment (kolesnikovae) [inline component/pyroscope/scrape/target.go:411]: Should we make agent to understand `pyroscope.io/application-name` annotation (see [example](https://github.com/grafana/pyroscope/tree/main/examples/golang-pull/kubernetes)) for backward compatibility with the existing deployments profiled with pyroscope OSS?

A nit: we could allocate the slice globally

Comment (korniltsev) [reply component/pyroscope/scrape/target.go:411]: I think if we try to support pyroscope.io annotations, we should support all of them, not just one and supporting all of the may complicate logic

Comment (Rperry2174) [reply component/pyroscope/scrape/target.go]: lets definitely not support all of them. I think if we support any we should support only the applicaiton name, but I'm also fine not supporting this and when we release pyroscope 1.0 we can have people make the change

Comment (eh-am) [reply component/pyroscope/scrape/target.go]: Plus not sure about using the `pyroscope.io` prefix.

File: docs/sources/flow/reference/components/pyroscope.scrape.md
Comment (kolesnikovae) [inline docs/sources/flow/reference/components/pyroscope.scrape.md:310]: We also should keep in mind that as long as we don't support removal from the db, it's might be quite frustrating experience for users to have "unknown" service in the dropdown (same with high cardinality values, such as pod name) – we should be ready to answer this question. On the other hand, provided that we fetch labels only from the head block, the "unknown" app will eventually disappear from the list once the user configuration is fixed.

File: component/pyroscope/scrape/target.go
Comment (kolesnikovae) [inline component/pyroscope/scrape/target.go:405]: The way I see it: a diligent user is supposed to create a relabeling config to populate `service_name` from an annotation/label or put it somehow else. If we provide a way to avoid this extra config by putting `pyroscope.io/name` label (for consistency, may be we should call it `pyroscope.io/service-name`), it would be nice to have it mentioned in the documentation.

**Review by kolesnikovae — APPROVED** [review]

File: docs/sources/flow/reference/components/pyroscope.scrape.md
Comment (Rperry2174) [reply docs/sources/flow/reference/components/pyroscope.scrape.md:310]: just trying to understand what this will look like in the UI:
image

cc @eh-am

File: component/pyroscope/scrape/target.go
Comment (kolesnikovae) [reply component/pyroscope/scrape/target.go]: My idea is to make the configuration process less prone to human errors. I expect that a good share of users that deal with this configuration already have experience of configuring Pyroscope OSS, where we have the `application-name` annotation. I anticipate that due to the human factor, these two might be mixed mistakenly.

File: docs/sources/flow/reference/components/pyroscope.scrape.md
Comment (eh-am) [reply docs/sources/flow/reference/components/pyroscope.scrape.md:310]: If it defaults to `unknown_service_name`, that's how it will look like in the UI, yes.

Differently from Anton, I feel like it's fine to have `unknown` or even fail to scrape altogether. IMO it's better to fail than trying to be too clever. When one is first setting up it's expected to have a few hiccups. As long as we document (and have proper error messages) it should be fine.

>On the other hand, provided that we fetch labels only from the head block, the "unknown" app will eventually disappear from the list once the user configuration is fixed.
But my understanding is that this is going to change sometime in the future, no?

File: component/pyroscope/scrape/target.go
Comment (Rperry2174) [reply component/pyroscope/scrape/target.go:405]: lets go with the concatenation (and risk the repetitions)

Comment (kolesnikovae) [reply component/pyroscope/scrape/target.go:405]: Yeah, that works. But I wouldn't defer to `pod_container_name` and `pod_name` labels even as a last resort. Moreover, I'd question necessity of falling back to "unknown" service. Maybe we should give up early.

---

On a side note: k8s labels vs annotations – I think it may be slightly better to use annotations only for our purposes. I understand that not everyone follows the rules, but still.

[Annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/): 
> You can use Kubernetes annotations to attach arbitrary non-identifying metadata to [objects](https://kubernetes.io/docs/concepts/overview/working-with-objects/#kubernetes-objects). Clients such as tools and libraries can retrieve this metadata.

[Labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/):
> Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users, but do not directly imply semantics to the core system.

Comment (Rperry2174) [reply component/pyroscope/scrape/target.go:405]: if "give up early" what happens from a user perspective?

File: docs/sources/flow/reference/components/pyroscope.scrape.md
Comment (kolesnikovae) [reply docs/sources/flow/reference/components/pyroscope.scrape.md:310]: > Differently from Anton, I feel like it's fine to have unknown or even fail to scrape altogether. IMO it's better to fail than trying to be too clever. When one is first setting up it's expected to have a few hiccups. As long as we document (and have proper error messages) it should be fine.

Just to clarify: I think that the label must be mandatory and explicit in both push and pull modes in the same way it is implemented in Pyroscope, and a missing service name should be threat as a misconfiguration. Profiles without the label should be discarded, and a corresponding message should be logged.

OTOH, I don't know how that works for ebpf.

Comment (Rperry2174) [reply docs/sources/flow/reference/components/pyroscope.scrape.md:310]: If it's not set it still seems like it would be easier to handle in the UI if there was some value that _means_ that it was not set (as opposed to not setting the label at all) wdyt @eh-am? would you rather have no `service_name` label or `service_name: undefined`

File: component/pyroscope/scrape/target.go
Comment (kolesnikovae) [reply component/pyroscope/scrape/target.go:405]: 1. No profiling data appears in the UI. So the user don't need to clean up after the unsuccessful attempt
2. A message indicating the misconfiguration is written to the agent log every time the target is discovered

Comment (korniltsev) [reply component/pyroscope/scrape/target.go:405]: I like the idea to give up early and not being clever for scrape.

For ebpf I still think we should go with concatenation or something else

File: component/pyroscope/scrape/target.go
Comment (rfratto) [inline component/pyroscope/scrape/target.go:277]: Is there a reason this is being removed? Isn't an `__address__` label still required so pyroscope.scrape knows where to collect profiles from?

File: component/pyroscope/scrape/target.go
Comment (eh-am) [reply component/pyroscope/scrape/target.go:405]: RE: eBPF concatenation, I don't think we need `namespace`, since it already be there (if you can infer via PID -> service discovery), right?

And based on the same principle, you can also look up the same tags you are looking for normal scrape (service_name, name etc).

Then as a last resort, the binary name could be used (somebody suggested it before), no?

File: component/pyroscope/scrape/target.go
Comment (rfratto) [reply component/pyroscope/scrape/target.go:277]: Ah, sorry, I missed that. Thanks!

File: component/pyroscope/scrape/target.go
Comment (Rperry2174) [reply component/pyroscope/scrape/target.go:235]: yeah confirming we decided on `service_name` and _not_ `__service_name__` so that this label is queryable

**Comment by korniltsev** [comment]
Now that we made `service_name` a required label it became another breaking change.
I suppose there will be changes required to be done in deployment_tools. Should I go ahead and prepare PR for deployment_tools or this code is owned by grafana agent team and I should not touch it?
@rfratto @cyriltovena

**Comment by rfratto** [comment]
> I suppose there will be changes required to be done in deployment_tools. Should I go ahead and prepare PR for deployment_tools or this code is owned by grafana agent team and I should not touch it?

You're more than free to touch that code, it's in a weird shared ownership state right now :)

File: docs/sources/flow/upgrade-guide.md
Comment (clayton-cornell) [inline docs/sources/flow/upgrade-guide.md:59]: Is there a word missing in this heading?

Comment (korniltsev) [reply docs/sources/flow/upgrade-guide.md:59]: thanks! fixed, made it 
```
`service_name` label in is required in `pyroscope.scrape`
```

File: component/pyroscope/scrape/target.go
Comment (kolesnikovae) [inline component/pyroscope/scrape/target.go:323]:
```suggestion
			return nil, nil, errors.New("service_name label is not specified")
```

**Review by clayton-cornell — COMMENTED** [review]
Looks OK for the docs.

File: docs/sources/flow/reference/components/pyroscope.scrape.md
Comment (Rperry2174) [reply docs/sources/flow/reference/components/pyroscope.scrape.md]: after discussing further in slack we decided that `service_name:undefined` may be better for the frontend, but will make querying significantly slower (compared to having an explicit service name that querying can rely on).

File: component/pyroscope/scrape/target.go
Comment (petethepig) [inline component/pyroscope/scrape/target.go:323]: same comment as [here](https://github.com/grafana/phlare/pull/745/files#diff-6ed840df23b67314462615ca210b760563632f290c2bd683c6dc716b1c622dc4)

**Comment by korniltsev** [comment]
I guess we need to decide on https://github.com/grafana/phlare/issues/751 before merging

**Comment by korniltsev** [comment]
@rfratto can we merge this?

**Pull request merged**
